src/Conversion/FileConversionManager.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace SilverStripe\Assets\Conversion;
+
+use SilverStripe\Assets\Storage\DBFile;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Injector\Injector;
+
+class FileConversionManager
+{
+    use Configurable;
+
+    private static $converters = [];
```

Comment:
1. The easiest way to arrange this is just a regular array of converters - and you just dump all of the converters in here and loop through all of them until you either find one that supports the conversion or run out of converters.
This is fine when there are only a few converters, assuming the logic in `supportsConversion()` doesn't have to do any serious computation.
This will _not_ scale well if there ends up being a lot of converters.
1. An alternative would be to register based on the category of file that the converter either converts _from_ or _to_, e.g:
    ```yml
    SilverStripe\Assets\Conversion\FileConversionManager:
      converters:
        image:
          SilverStripe\Assets\Conversion\InterventionImageConverter
          App\Conversion\PdfToImageConverter
        image/supported:
          SilverStripe\Assets\Conversion\InterventionImageConverter
          App\Conversion\PdfToImageConverter
        document:
          App\Conversion\WordToPdfConverter
    ```
    - The above uses the category names from `File.app_categories` configuration - but note that image converters need to be doubled up, because there's two separate image categories.
    - Note it's also a) not clear whether these are what the converter supports converting _to_ or _from_ (e.g. `PdfToImageConverter` is in the `image` and `image_supported` categories because it converts _to_ those - but developers could easily put it in the `document` category accidentally thinking they're meant to put them in the category they convert _from_ (or vice versa)
1. Similar to 2 above, except they could be listed everywhere they convert *from* and everywhere they convert *to*. This would ensure we only ever check converters that convert from-and-to the requested formats - but it's even more confusing to configure. I would definitely not recommend this unless we find people are making thousands of converters.
1. Similar to 3 above, but flip the config on its head like so:
    ```yml
    SilverStripe\Assets\Conversion\FileConversionManager:
      converters:
          SilverStripe\Assets\Conversion\InterventionImageConverter:
            from: ['image', 'image/supported']
            to: ['image', 'image/supported']
          App\Conversion\PdfToImageConverter:
            from: 'document'
            to: ['image', 'image/supported']
          App\Conversion\WordToPdfConverter
    ```
    - Any converter with no listed "from" or "to" will either throw an exception for bad config, or more likely just always check those, but check them after everything else.
    - Any converter which _does_ list a "from" and a "to" category can be skipped early if the format to convert from-or-to isn't in their list.
1. ??? I'm sure there are other ways we could organise this - and if someone has a great idea I'd be happy to hear it.

Right now I'm inclined to go with option 3 - I think it has a nice trade-off between being easy to understand and outright skipping converters we know won't handle conversions without even asking.

Reply:
I would lean toward option 1. I think the entire point of converters is to make this process as easy and transparent as possible for the end developer. 

If someone has very difficult and complex needs, that's the point where maybe you want to use the lower level API directly instead of using a converter.

Maybe we need a way to apply priorities so people can override converters or force their converter to take precedence over another.

Reply:
> If someone has very difficult and complex needs, that's the point where maybe you want to use the lower level API directly instead of using a converter.

I think this misses the point a bit - it's just about avoiding asking converters if they can handle a conversion if the conversion is way out of its scope. e.g. don't ask an image converter if it can convert .doc to .pdf.

> Maybe we need a way to apply priorities so people can override converters or force their converter to take precedence over another.

Overriding converters should be done via the injector.
Precedence does make some sense though. Someone might have a hyper-efficient converter for some specific format (for example converting to webp) - but which don't handle anything else. It would make sense to want that to come before any other image converters.

I'll see what I can do in that regard, I think that makes a lot of sense.

Reply:
Actually nevermind - priority isn't needed. People should just use the `After` and `Before` configuration directives.

Reply:
For future reference: Since both Max and I leaned towards option 1 that's what I'll go with.